source/Nevermore/Util/DataModificationQueryBuilder.cs

```diff
@@ -330,87 +329,155 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, string> allocateI
             return result;
         }
 
-
-        void AppendRelatedDocumentStatementsForInsert(
-            StringBuilder sb,
-            CommandParameterValues parameters,
+        PreparedCommand[] CreateInsertCommandsWithRelatedDocuments(
+            StringBuilder insertStatementBuilder,
             DocumentMap mapping,
-            IReadOnlyList<object> documents)
+            IReadOnlyList<object> documents,
+            InsertOptions options)
         {
+            var commands = new List<PreparedCommand>();
+            var parameters = new CommandParameterValues();
+            // we need to create these document parameters for each command - parameters are disposed at the end of execution in 'CommandExecutor'
+            Func<CommandParameterValues> documentParametersCreator = () => GetDocumentParameters(m => keyAllocator(m), options.CustomAssignedId, documents, mapping, DataModification.Insert);
+            parameters.AddRange(documentParametersCreator());
+
             var relatedDocumentData = GetRelatedDocumentTableData(mapping, documents);
+            if (!relatedDocumentData.Any(x => x.Related.Any()))
```

Comment:
`!relatedDocumentData.Any(x => x.Related.Any()` => `if (!RelatedDocumentDataPresent)` ?. It's used in a few places and this reads better for me